lib/Word_Lib/More_Bit_Ring.thy

```diff
@@ -0,0 +1,131 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+theory More_Bit_Ring
+  imports Main
+begin
+
+(* Additional equalities on semiring_bit_operations and ring_bit_operations, in particular
+   relationships between Boolean and arithmetic operations. *)
+
+context semiring_bit_operations
+begin
+
+context
+  includes bit_operations_syntax
+begin
+
+lemma disjunctive_add2:
+  "(x AND y) = 0 \<Longrightarrow> x + y = x OR y"
+  by (metis disjunctive_add bit_0_eq bit_and_iff bot_apply bot_bool_def)
+
+end
+end
+
+context ring_bit_operations
+begin
+
+context
+  includes bit_operations_syntax
+begin
+
+lemma not_xor_is_eqv:
+  "NOT (x XOR y) = (x AND y) OR (NOT x AND NOT y)"
+  by (simp add: bit.xor_def bit.disj_conj_distrib or.commute)
+
+lemma not_xor_eq_xor_not:
+  "(NOT x) XOR y = x XOR (NOT y)"
+  by simp
+
+lemma not_minus:
+  "NOT (x - y) = y - x - 1"
+  by (simp add: not_eq_complement)
+
+lemma minus_not_eq_plus_1:
+  "- NOT x = x + 1"
+  by (simp add: minus_eq_not_plus_1)
+
+lemma not_minus_eq_minus_1:
+  "NOT (- x) = x - 1"
+  by (simp add: not_eq_complement)
+
+lemma and_plus_not_and:
+  "(x AND y) + (x AND NOT y) = x"
+  by (metis and.left_commute and.right_neutral bit.conj_cancel_right bit.conj_disj_distrib
+            bit.conj_zero_right bit.disj_cancel_right disjunctive_add2)
+
+lemma or_eq_and_not_plus:
+  "x OR y = (x AND NOT y) + y"
+  by (simp add: and.assoc bit.disj_conj_distrib2 disjunctive_add2)
+
+lemma and_eq_not_or_minus:
+  "x AND y = (NOT x OR y) - NOT x"
+  by (metis and.idem and_eq_not_not_or eq_diff_eq or.commute or.idem or_eq_and_not_plus)
+
+lemma and_not_eq_or_minus:
+  "x AND NOT y = (x OR y) - y"
+  by (simp add: or_eq_and_not_plus)
+
+lemma and_not_eq_minus_and:
+  "x AND NOT y = x - (x AND y)"
+  by (simp add: add.commute eq_diff_eq and_plus_not_and)
+
+lemma or_minus_eq_minus_and:
+  "(x OR y) - y = x - (x AND y)"
+  by (metis and_not_eq_minus_and and_not_eq_or_minus)
+
+lemma plus_eq_and_or:
+  "x + y = (x OR y) + (x AND y)"
+  using add_commute local.add.semigroup_axioms or_eq_and_not_plus semigroup.assoc
+  by (fastforce simp: and_plus_not_and)
+
+lemma xor_eq_or_minus_and:
+  "x XOR y = (x OR y) - (x AND y)"
+  by (metis (no_types) bit.de_Morgan_conj bit.xor_def2 bit_and_iff bit_or_iff disjunctive_diff)
+
+lemma nxor_eq_and_plus_not_or:
```

Comment:
nxnor? not_xnor would fit better with the names of the others, and xor lemmas aren't going to be in common use anyway

Reply:
you mean `not_xor`? The thing is that `not xor` is the same as `equals` (usually written `≡`), and I almost defined that as an operator, but then we'd have to have all kinds of basic rules and setup for it and the symbol is already taken etc, so I left it, but I still had the feeling that `not_xor` is a long name for a thing that is actually simpler than `xor`.

But you're right, `nxor` doesn't really make anything better. Will change to `not_xor`.